docs/how-to-guides/config-settings.rst

```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
 
 .. _build: https://pypa-build.readthedocs.io/en/stable/
 .. _pip: https://pip.pypa.io/
+
+
+Using a persistent build directory
+==================================
+
+By default, ``meson-python`` will create a temporary build directory, which
+gets cleaned up after the sdist or wheel has been built, or if the build fails.
+The ``build-dir`` config seting (see :ref:`reference-config-settings`) can be
+used to build in a user-specified build directory instead. For example:
+
+.. tab-set::
+
+    .. tab-item:: pypa/build
+        :sync: key_pypa_build
+
+	.. code-block:: console
+
+	   $ python -m build -Cbuild-dir=build
+
+    .. tab-item:: pip
+        :sync: key_pip
+
+	.. code-block:: console
+
+	   $ python -m pip install . -Cbuild-dir=build
+
+
+The ``build`` directory will now contain all the build output created by
+``meson``, ``ninja`` and ``meson-python``. This may be useful for a couple of
+reasons:
+
+- For development efficiency (incremental builds)
+- For accessing build logs and intermediate build outputs when debugging
+
+The same build directory can be used by subsequent invocations of
+``meson-python``. This avoids having to rebuild the whole project when testing
+changes during development. This will work with any config settings flags,
+including for cross compilation. The one thing to keep in mind is that changing
+the version of ``meson`` or ``meson-python`` in the build environment between
+separate builds in the same build directory is not supported (it often works,
+but may fail - and the error message cannot be guaranteed to be clear about the
+reason).
```

Comment:
I don't think this is the case anymore. We fixed meson-python so that `--reconfigure` is used when needed.

Reply:
I'm not sure that `meson setup --reconfigure` is 100% the same as `rm -rf builddir && meson setup builddir` though. I recently ran into a very odd failure (https://github.com/scipy/scipy/issues/20433) which had to do with a pickled version of `meson` that is reloaded at build time.

The Meson docs don't say anything about this either way I believe, but my take-away from https://github.com/mesonbuild/meson/pull/12357#issuecomment-2050016590 was "if you upgrade Meson, no guarantees unless you delete the build dir". I could be wrong though - @eli-schwartz?

Reply:
Reconfigure is *definitely* a section of the codebase covered by meson's explicit check for coredata version.

If meson detects that you've upgraded meson and the version of the coredata doesn't match, it essentially does a wipe and from-scratch configure, instead of the normal reconfigure logic. Alternatively, if you haven't asked to setup the project it isn't apropos to do so automatically just because the coredata is incompatible -- from memory, things like test/install will just print a "fatal error: xxx. Try running meson setup --wipe".

Reconfigure is indeed different from deleting the build directory. It will:
- keep dependency lookup caches instead of actually running pkg-config/cmake for dependency()
- keep incremental build outputs, since they will anyways be invalidated if the command line changes
- preserve current option states

Reply:
thanks both. removed these sentences now.